src/main/java/org/araymond/joal/core/client/emulated/generator/key/TimedOrAfterStartedAnnounceRefreshKeyGenerator.java

```diff
@@ -4,58 +4,41 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.annotations.VisibleForTesting;
 import com.turn.ttorrent.common.protocol.TrackerMessage.AnnounceRequestMessage.RequestEvent;
+import lombok.Getter;
 import org.araymond.joal.core.client.emulated.TorrentClientConfigIntegrityException;
 import org.araymond.joal.core.client.emulated.generator.key.algorithm.KeyAlgorithm;
 import org.araymond.joal.core.client.emulated.utils.Casing;
 import org.araymond.joal.core.torrent.torrent.InfoHash;
 
 import java.time.LocalDateTime;
-import java.time.temporal.ChronoUnit;
 
 /**
  * Created by raymo on 16/07/2017.
  */
-public class TimedOrAfterStartedAnnounceRefreshKeyGenerator extends KeyGenerator {
-    @VisibleForTesting
-    LocalDateTime lastGeneration;
-    private String key;
-    private final Integer refreshEvery;
+public class TimedOrAfterStartedAnnounceRefreshKeyGenerator extends TimedRefreshKeyGenerator {
 
     @JsonCreator
     TimedOrAfterStartedAnnounceRefreshKeyGenerator(
             @JsonProperty(value = "refreshEvery", required = true) final Integer refreshEvery,
             @JsonProperty(value = "algorithm", required = true) final KeyAlgorithm algorithm,
             @JsonProperty(value = "keyCase", required = true) final Casing keyCase
     ) {
-        super(algorithm, keyCase);
-        if (refreshEvery == null || refreshEvery < 1) {
-            throw new TorrentClientConfigIntegrityException("refreshEvery must be greater than 0.");
-        }
-        this.refreshEvery = refreshEvery;
-    }
-
-    @JsonProperty("refreshEvery")
-    Integer getRefreshEvery() {
-        return refreshEvery;
+        super(refreshEvery, algorithm, keyCase);
     }
 
     @Override
     public String getKey(final InfoHash infoHash, final RequestEvent event) {
-        if (this.shouldRegenerateKey()) {
+        if (super.shouldRegenerateKey()) {
             this.lastGeneration = LocalDateTime.now();
-            this.key = super.generateKey();
+            setKey(generateKey());
         }
 
```

Comment:
Those 4 lines are useless because the time logic is already applied by `TimedRefreshKeyGenerator#getKey()`, which is called on line 36

Reply:
I have to admit my eyes glossed over a bit around the key generators.

Not sure you are right here, as the generation on #36 is only called if `event = STARTED`.
But something is surely wrong there; eg the return var `key`, initialized on #33, is not updated if new key is generated on #36.

Should the method maybe look like
```
    public String getKey(final InfoHash infoHash, final RequestEvent event) {
        if (event == RequestEvent.STARTED || super.shouldRegenerateKey()) {
            this.lastGeneration = LocalDateTime.now();
            setKey(generateKey());
        }

        return getKey();
    }
```
?

But again, I don't fully comprehend the the key generation part of the codebase.

Reply:
As for 
> the time logic is already applied by TimedRefreshKeyGenerator#getKey(), which is called on line 36

then that's not quite right - `TimedRefreshKeyGenerator#getKey()` is lombok-generated simple field getter, ie that method doesn't have any logic.

Reply:
woops, i've confused `super.getKey()` which does not apply refresh logic and `super.getKey(infoHash, event)` which does.
Indeed i like the better the second approach.

    public String getKey(final InfoHash infoHash, final RequestEvent event) {
        if (event == RequestEvent.STARTED || super.shouldRegenerateKey()) {
            this.lastGeneration = LocalDateTime.now();
            setKey(generateKey());
        }

        return getKey();
    }